src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -3,75 +3,64 @@ package mesosphere.marathon.state
 import com.wix.accord.dsl._
 import com.wix.accord._
 import mesosphere.marathon.api.v2.Validation._
-import org.apache.mesos.{ Protos => Mesos }
+
+import org.apache.mesos.Protos.ContainerInfo
+
 import scala.collection.immutable.Seq
 
-// TODO: trait Container and specializations?
-// Current implementation with type defaulting to DOCKER and docker to NONE makes no sense
-case class Container(
-    `type`: Mesos.ContainerInfo.Type = Mesos.ContainerInfo.Type.DOCKER,
-    volumes: Seq[Volume] = Nil,
-    docker: Option[Container.Docker] = None) {
+trait Container {
+  val volumes: Seq[Volume]
+
+  def forDocker(f: Container.Docker => Unit): Unit = {
+    this match {
+      case docker: Container.Docker => f(docker)
+      case _ =>
+    }
+  }
+
+  def foldDocker[R](defaultResult: R)(f: Container.Docker => R): R = {
+    this match {
+      case docker: Container.Docker => f(docker)
+      case _ => defaultResult
+    }
+  }
+
+  def mapDocker(f: Container.Docker => Container.Docker): Container = {
```

Comment:
Oops, leftovers from the other solution. Removing the two unused functions (keeping `foldDocker` only. (BTW: this kind of folding of one select case, not just for Option, should be a general feature of the language. Any tips where I can find such a thing?)